src/modules/sovereign-clouds/index.ts

```diff
@@ -0,0 +1,55 @@
+import { AUTH_URL, GRAPH_URL } from '../../app/services/graph-constants';
+import { geLocale } from '../../appLocale';
+import { ICloud } from '../../types/cloud';
+
+const storageKey = 'cloud';
+
+export const clouds: ICloud[] = [
+  {
+    locale: 'en-us',
+    name: 'China',
+    baseUrl: 'https://microsoftgraph.chinacloudapi.cn',
+    loginUrl: 'https://login.chinacloudapi.cn'
+  },
+  {
+    locale: 'global',
+    name: 'Canary',
+    baseUrl: 'https://canary.graph.microsoft.com',
+    loginUrl: 'https://login.microsoftonline.com'
+  }
+];
+
+export const globalCloud: ICloud = {
+  baseUrl: GRAPH_URL,
+  locale: 'global',
+  loginUrl: AUTH_URL,
+  name: 'Global'
+}
+
+export function getCurrentCloud(): ICloud | undefined {
+  const cloudName = localStorage.getItem(storageKey);
+  if (cloudName && cloudName === globalCloud.name) {
+    return globalCloud;
+  }
+  return (cloudName) ? getCloudProperties(cloudName) : undefined;
+}
+
+export function getEligibleCloud(): ICloud | undefined {
+  const localeValue = geLocale.toLowerCase();
+  return clouds.find(k => k.locale === localeValue);
+}
+
+export function getCloudProperties(cloudName: string): ICloud | undefined {
+  return clouds.find(k => k.name === cloudName) || undefined;
+}
+
+export function replaceBaseUrl(url: string): string {
+  const { origin } = new URL(url);
+  const currentCloud = getCurrentCloud() || null;
```

Comment:
I used `|| null` to help to default the undefined response to a null so that I can perform a null check.
I noticed that trying `if (currentCloud) { ... }` would not work when I get undefined.

Reply:
Just changed this to return null instead